### PR TITLE
sln-add: Don't add solution folders outside the scope of solution directory

### LINKED ISSUE #48608: dotnet sln add Doesn't Work on Cross-Drive Path
### Describe the bug
In Dev Kit, we create a solution file in VSCode's workspace folder when users enable "solution-free" mode. When users create projects, they are located in the user's open workspace. My solution file is located on my C: drive, while the project is located on my Q: drive. The command that I would run is:

`dotnet sln "c:\\Users\\xxx\\Code\\User\\workspaceStorage\\ceea826104b74110846cf9474b011fd6\\ms-dotnettools.csdevkit\\ConsoleApp2.sln" add "q:\\src\\source\\ConsoleApp2\\bar\\bar.csproj`

However, I encounter the following issue:

```
Names cannot:
- contain any of the following characters: / ? : \ * " < > |
- contain control characters
- be system reserved names, including 'CON', 'AUX', 'PRN', 'COM1' or 'LPT2'
- be '.' or '..' (Parameter 'name')
```

### To Reproduce

1. Create a solution file in the C: drive.
2. Create a project in the Q: drive.
3. Use the dotnet sln add command to add the project to the solution file.
4. Observe the error message.

### Further technical details
- Include the output of `dotnet --info`
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version

`dotnet info`
                            
.NET SDK:                                                                
 Version:           9.0.203                                              
 Commit:            dc7acfa194
 Workload version:  9.0.200-manifests.9df47798
 MSBuild version:   17.13.20+a4ef1e90f

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.26100
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.203\

.NET workloads installed:
There are no installed workloads to display.
Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.4
  Architecture: x64
  Commit:       f57e6dc747

.NET SDKs installed:
  8.0.400 [C:\Program Files\dotnet\sdk]
  8.0.408 [C:\Program Files\dotnet\sdk]
  9.0.105 [C:\Program Files\dotnet\sdk]
  9.0.203 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

### PR DESCRIPTION
**Pull request by edvilme**
Fixes #48608 

# Description
When adding a project to a solution file, solution folders are automatically generated with the relative path (relative to the solution file path). When adding projects outside the scope of the solution working directory, solution folders with invalid names are created (e.g., `..`, `:`, etc).

This was addressed in #46456 but didn't consider paths in different volumes. 

# Risk
Low - this doesn't alter existing behavior but rather considers a previously not considered edge case.

# Regression 
Yes - this worked before 9.0.2xx

# Testing
Adding automated tests is difficult as it involves mounting several volumes, but manual tests were performed.

 ## Expected behavior
image

## Current behavior
image

### CODE REVIEW

File: src/Cli/dotnet/commands/dotnet-sln/add/Program.cs [hunk 1]
```diff
@@ -100,9 +102,9 @@ private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, IEnum
             foreach (var projectPath in projectPaths)
             {
                 string relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
-                // Add fallback solution folder if relative path does not contain `..`.
-                string relativeSolutionFolder =  relativePath.Split(Path.DirectorySeparatorChar).Any(p => p == "..")
-                    ? string.Empty : Path.GetDirectoryName(relativePath);
+
+                // Add fallback solution folder if relative path does not contain invalid characters
+                string relativeSolutionFolder = DoesSolutionFolderPathContainInvalidChars(relativePath) ? string.Empty : Path.GetDirectoryName(relativePath);
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:107]: I got turned around twice reading this. I think I understand it properly now. I feel like there has to be a better way to express this, but I haven't figured one out yet, to be honest

Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:107]: Yes, It is very messy. I am refactoring the whole "automatically generate intermediate solution folders" part to be more clear on what it should do

File: src/Cli/dotnet/commands/dotnet-sln/add/Program.cs [hunk 2]
```diff
@@ -31,6 +31,13 @@ private static string GetSolutionFolderPathWithForwardSlashes(string path)
             return "/" + string.Join("/", PathUtility.GetPathWithDirectorySeparator(path).Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries)) + "/";
         }
 
+        private static bool DoesSolutionFolderPathContainInvalidChars(string pathFragment)
+        {
+            return pathFragment
+                .Split(Path.DirectorySeparatorChar)
+                .Any(pathFragment => pathFragment == ".." || pathFragment == "." || Path.GetInvalidFileNameChars().Any(c => pathFragment.Contains(c)));
```
[the remark this answers is not in the document]
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:39]: This should be
```suggestion
                .Any(pathFragment => pathFragment == ".." || pathFragment == "." || Path.GetInvalidFileNameChars().Any(c => pathFragment.Equals(c)));
```

What we are trying to do here is see if any of the fragments _is_ an invalid char, i.e., https://github.com/dotnet/runtime/blob/2224669d22aa1f4dd2aa953942ba8945aabfa9e9/src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs#L15

[the remark this answers is not in the document]
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:39]: You're right, I thought it should check to see if individual components _are_ the unvalid strings, but yes, it should check if the path contains them

File: src/Cli/dotnet/commands/dotnet-sln/add/Program.cs [hunk 3]
```diff
@@ -31,13 +32,21 @@ private static string GetSolutionFolderPathWithForwardSlashes(string path)
             return "/" + string.Join("/", PathUtility.GetPathWithDirectorySeparator(path).Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries)) + "/";
         }
 
+        private static bool IsSolutionFolderPathInDirectoryScope(string relativePath)
+        {
+            return !string.IsNullOrEmpty(relativePath)
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:37]: So for instance, if I had a path like "src\Build\..\..\..\..\test\foo"...

File: src/Cli/dotnet/commands/dotnet-sln/add/Program.cs [hunk 4]
```diff
@@ -82,15 +89,10 @@ private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, IEnum
...
-                foreach (var platform in _defaultPlatforms)
-                {
-                    solution.AddPlatform(platform);
-                }
-                foreach (var buildType in _defaultBuildTypes)
-                {
-                    solution.AddBuildType(buildType);
-                }
+                _defaultPlatforms.ToList().ForEach(solution.AddPlatform);
+                _defaultBuildTypes.ToList().ForEach(solution.AddBuildType);
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs:96]: It's also a random perf hit and less "standard." I prefer the original, personally

**Review by Forgind — COMMENTED** [review]
Though I generally look favorably upon trying to clean up code, I don't think it's positive in servicing PRs, as it makes it harder to review and can lead to unexpected errors. I found GenerateIntermediateSolutionFoldersForProjectPath particularly confusing. I'm not saying this can't go in, but I don't think it should go into 9.0.3xx as-is.

**Review by Forgind — APPROVED** [review]

**Comment by Forgind** [comment]
I think you could potentially be more lenient with things with .. that end up still in the sln's cone, but 🤷

File: src/Cli/dotnet/commands/dotnet-sln/add/Program.cs (hunk 3 above)
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs]: I am not sure it is entirely possible to run into this issue since `relativePath` is obtained through `Path.GetRelativePath`, which should return the actual resolved path.
So instead of `src\Build\..\..\..\..\test\foo`, it should be `..\..\test\foo`

Comment (Forgind) [reply src/Cli/dotnet/commands/dotnet-sln/add/Program.cs]: Fair enough; missed that
